protocols/raft/src/main/java/io/atomix/protocols/raft/storage/snapshot/SnapshotFile.java

```diff
@@ -90,6 +93,19 @@ static File createSnapshotFile(File directory, String serverName, long index) {
     return new File(directory, createSnapshotFileName(serverName, index));
   }
 
+  /**
+   * Creates a temporary file for writing snapshots.
+   */
+  static File createTemporaryFile(File base) {
+    try {
+      final File file = File.createTempFile(base.getName(), null);
+      file.deleteOnExit();
```

Comment:
@npepinpe please replace this method call with http://docs.oracle.com/javase/7/docs/api/java/nio/file/StandardOpenOption.html#DELETE_ON_CLOSE to avoid possible memory leak with shutdown hooks.

Reply:
Or better yet can we just delete it when done with it? There's a few SO threads where they say the file is deleted immediately on some systems.

Reply:
The `deleteOnExit` here is just a failsafe in case the snapshot isn't closed; the temp file is removed when closing the snapshot or completing it. `DELETE_ON_CLOSE` seems more robust than `File.deleteOnExit()` though, so I'll update it to that.

> Or better yet can we just delete it when done with it? There's a few SO threads where they say the file is deleted immediately on some systems.

Do you mean that opening it with `DELETE_ON_CLOSE` causes it to delete the file prematurely?

Reply:
Might be a little easier for you to verify but is the shutdown hook removed if the file is closed?

Reply:
Good catch, it's definitely not removed. `DELETE_ON_CLOSE` would be preferable (though I'm not sure how it's implemented). But I think using `DELETE_ON_CLOSE` means refactoring `FileBytes` to use NIO channels instead of `RandomAccessFile`. In either case I think `deleteOnExit` is not really all that useful then.

Reply:
Might be easier to just make a best effort that the file is deleted when we are done with it then.

Reply:
The file is deleted when the snapshot is closed or completed. Snapshots loaded from disks do not have a temp file associated, so they also leave nothing behind. `deleteOnClose` in this case is only useful if we introduce an error where we forget to close or complete snapshots, since it doesn't work if the JVM crashes. I could take it or leave it (leaning more towards leaving it imo), up to you.